pkg/controller/clusterdeployment/clusterinstalls.go

```diff
@@ -0,0 +1,239 @@
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	hivecontractsv1alpha1 "github.com/openshift/hive/apis/hivecontracts/v1alpha1"
+	"github.com/openshift/hive/pkg/constants"
+	hivemetrics "github.com/openshift/hive/pkg/controller/metrics"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+)
+
+func (r *ReconcileClusterDeployment) reconcileExistingClusterInstall(cd *hivev1.ClusterDeployment, logger log.FieldLogger) (reconcile.Result, error) {
+	ref := cd.Spec.ClusterInstallRef
+	gvk := schema.GroupVersionKind{
+		Group:   ref.Group,
+		Version: ref.Version,
+		Kind:    ref.Kind,
+	}
+
+	logger = logger.WithField("clusterinstall", ref.Name).WithField("gvk", gvk)
+	logger.Debug("reconciling existing clusterinstall")
+
+	ci := &hivecontractsv1alpha1.ClusterInstall{}
+	err := controllerutils.GetDuckType(context.TODO(), r.Client,
+		gvk,
+		types.NamespacedName{Namespace: cd.Namespace, Name: ref.Name},
+		ci)
+	if apierrors.IsNotFound(err) {
+		logger.Debug("cluster is not found, so skipping")
+		return reconcile.Result{}, nil
+	}
+	if err != nil {
+		logger.WithError(err).Error("failed to get the cluster install")
+		return reconcile.Result{}, err
+	}
+
+	specModified := false
+	statusModified := false
+	// copy the cluster metadata
+	if met := ci.Spec.ClusterMetadata; met != nil {
+		if met.InfraID == "" ||
+			met.ClusterID == "" ||
+			met.AdminKubeconfigSecretRef.Name == "" ||
+			met.AdminPasswordSecretRef.Name == "" {
+			return reconcile.Result{}, nil // the information is not complete and therefore skip
+		}
+		if !reflect.DeepEqual(cd.Spec.ClusterMetadata, ci.Spec.ClusterMetadata) {
+			cd.Spec.ClusterMetadata = ci.Spec.ClusterMetadata
+			specModified = true
+		}
+	}
+
+	if cd.Status.InstallRestarts != ci.Status.InstallRestarts {
+		cd.Status.InstallRestarts = ci.Status.InstallRestarts
+		statusModified = true
+	}
+
+	// copy the required conditions
+	requiredConditions := []hivecontractsv1alpha1.ClusterInstallConditionType{
+		hivecontractsv1alpha1.ClusterInstallFailed,
+		hivecontractsv1alpha1.ClusterInstallCompleted,
+		hivecontractsv1alpha1.ClusterInstallRequirementsMet,
+	}
+
+	conditions := cd.Status.Conditions
+	for _, req := range requiredConditions {
+		cond := controllerutils.FindClusterInstallCondition(ci.Status.Conditions, req)
+		if cond == nil {
+			continue
+		}
+
+		updated := false
+		conditions, updated = controllerutils.SetClusterDeploymentConditionWithChangeCheck(conditions,
+			hivev1.ClusterDeploymentConditionType("ClusterInstall"+string(cond.Type)), // this transformation is part of the contract
```

Comment:
Proposed cond.Type just being a string above.

Reply:
it is string technically, just the go type for our contract is specific for better handling of finding conditions for our controllers.

i'm okay if you want the clsuterinstall contract's conditions type to be string